### ISSUE TITLE
electrum unwanted disconnect from own ElectrumX server

### ISSUE THREAD

**Issue by adrelanos** [spesmilo/electrum#7549]
**Steps to reproduce:**

1) Setup ElectrumX.
2) Setup electrum with `oneserver` connecting one's own ElectrumX server that was just set up and is not fully blockchain synced yet.

**Expected result:**

electrum shows status "in progress" (yellow light), "not all transactions may display, electrum server behind bitcoin blockchain" or something equivalent, waits for one's own ElectrumX server to catch up with the Bitcoin blockchain download and notifies about that in its GUI

**Actual result:**

electrum shows status "disconnected" (red light). It is difficult to figure out what's wrong.

Electrum in verbose mode (`-v`) on the command line reports:

```
I/i | interface.[localhost:5000] | connection established. version: ['ElectrumX 1.16.0', '1.4']
I/i | interface.[localhost:5000] | disconnecting due to GracefulDisconnect('server tip below max checkpoint')
```

**Why is that an issue?**

Unexpected and very difficult to figure out what's wrong. One starts debugging.... "Does Bitcoin Core have an issue downloading the blockchain? No. Bitcoin RPC connection reachable? Yes. ElectrumX connecting to Bitcoin Core RPC? Yes. ElectrumX local server TCP port open and reachable? Yes. Electrum configured to connect to the right server (localhost)? Yes." So why does it not connect? Check electrum on command line in verbose mode.

This seems to be a bug at least in context of using electrum with one's own self-hosted, local ElectrumX server.

This might make sense when using electrum with public third party servers but it's very confusing when using electrum `oneserver` + own ElectrumX.

(Setting up one's own ElectrumX server and to connect to it with electrum is very difficult because upstream documentation for electrum and ElectrumX is lacking / more like developer documentation than an actionable guide that users can follow. Whonix or otherwise, would be equally difficult in other Linux distributions such as Debian. Therefore, [how to use ElectrumX in Whonix](https://www.whonix.org/wiki/ElectrumX) has just been documented by me. I am a developer of Whonix. Mentioning just in case you'd like to see full instructions for reproduction.)

**Suggested Solution**

* Could you please add an (command line configuration) option `trustedserver` (or better name)? If enabled, do not disconnect from the server.
* If using option electrum with option `trustedserver`, and ElectrumX is not fully synced yet, electrum shows status "in progress" (yellow light), "not all transactions may display, electrum server behind bitcoin blockchain" or something equivalent, waits for one's own ElectrumX server to catch up with the Bitcoin blockchain download and notifies about that in its GUI

**Related**

[ElectrumX should accept local RDP connections even if it has not fully caught up with the bitcoin daemon](https://github.com/spesmilo/electrumx/issues/161)

**Comment by benpbolton** [comment]
I don't have a disconnect problem with my test electrumx single server test setups. 

Which protocol are you using to connect to your electrumx server, or what's your full connection string (eg. `...  --oneserver --server localhost:50001:t`) 

I assume your electrumx.conf contains the appropriate services like `SERVICES = tcp://:50001,rpc://` ... what does the electrumx logging state when you're disonnected, eg. what's the output of `journalctl -f -u electrumx`

**Comment by adrelanos** [comment]
Is it really contested that this is the case?

I should have mentioned, it is written in documentation.

Quote https://electrumx-spesmilo.readthedocs.io/en/latest/environment.html

> ElectrumX will not serve any incoming connections until it has fully caught up with your bitcoin daemon.

I am arguing that this shouldn't be the case. Should be changed. This seems to be a bug at least in context of using electrum with one's own self-hosted, local ElectrumX server.

Therefore either I am wrong or at least the documentation is wrong?

**Comment by benpbolton** [comment]
Gotcha, I understand more fully now. It sounds like your issue is more with electrumx (and should be filed there, not in electrum?) as only supporting a local rcp connection when not synced to the tip of your node? 

As for electrum getting another switch, take a step back:
1) you think the documentation is obtuse or complex/complicated as it is
2) you recommend adding an additional switch for your unique use-case (likely further exacerbating the first point)

I could see an eventual support for... say... a yellow-light with hover-over "not all transactions may display, electrum server behind bitcoin blockchain" or something equivalent... but I can't say I'm in favor of a lying green light, when all might not be well with the 'network'

**Comment by adrelanos** [comment]
> It sounds like your issue is more with electrumx (and should be filed there, not in electrum?) as only supporting a local rcp connection when not synced to the tip of your node?

Indeed. I should have opened under ElectrumX, not electrumx. Can the ticket be moved or should I close and create a new, improved ticket?

> As for electrum getting another switch, take a step back:
> 
> 1. you think the documentation is obtuse or complex/complicated as it is
> 2. you recommend adding an additional switch for your unique use-case (likely further exacerbating the first point)

A switch could be avoided if connection is to localhost then it could be considered auto-detected?

> I could see an eventual support for... say... a yellow-light with hover-over "not all transactions may display, electrum server behind bitcoin blockchain" or something equivalent... but I can't say I'm in favor of a lying green light, when all might not be well with the 'network'

That sounds great!

**Comment by adrelanos** [comment]
There are actually two different issues. One for electrum and one for ElectrumX.

* electum issue, this issue in this ticket: `electrum unwanted disconnect from own ElectrumX server`
* ElectrumX issue: [`ElectrumX should accept local RDP connections even if it has not fully caught up with the bitcoin daemon`](https://github.com/spesmilo/electrumx/issues/161)

I've updated the original issue description in this ticket just now for better clarity after this discussion. Please kindly let me know in case it still needs improvements.
